File: java/src/test/java/money_problem/domain/properties/MoneyGenerator.java
package money_problem.domain.properties;

import com.pholser.junit.quickcheck.generator.GenerationStatus;
import com.pholser.junit.quickcheck.generator.Generator;
import com.pholser.junit.quickcheck.random.SourceOfRandomness;
import money_problem.domain.Currency;
import money_problem.domain.Money;

public class MoneyGenerator extends Generator<Money> {
    public static final int MAX_AMOUNT = 1_000_000_000;

    public MoneyGenerator() {
        super(Money.class);
    }

    @Override
    public Money generate(SourceOfRandomness sourceOfRandomness, GenerationStatus generationStatus) {
        return new Money(
                sourceOfRandomness.nextDouble(-MAX_AMOUNT, MAX_AMOUNT),
                sourceOfRandomness.choose(Currency.values())
        );
    }
}

File: c#/money-problem.Tests/MoneyGenerator.cs
using System;
using FsCheck;
using money_problem.Domain;

namespace money_problem.Tests;

public class MoneyGenerator
{
    private const double MaxAMount = 1000000000;
    private const int MaxDigits = 5;

    public static Arbitrary<Money> GenerateMoneys() =>
        Arb.From(from amount in GetAmountGenerator()
            from currency in Arb.Generate<Currency>()
            select new Money(amount, currency));

    private static Gen<double> GetAmountGenerator() => Arb.Default.Float()
        .MapFilter(x => Math.Round(x, MaxDigits), x => x is >= 0 - MaxAMount and <= MaxAMount).Generator;
}